torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1202,8 +1216,18 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
   ncclIdToCommMap_.emplace(buildNcclUniqueIdStr(ncclID), ncclComms);
 
   // Move the NCCL resource to cache
-  devNCCLCommMap_.emplace(devicesKey, std::move(ncclComms));
-  return devNCCLCommMap_[devicesKey];
+  auto it = inInitializationCommMap_.find(devicesKey);
+  // A previous thread could've already removed devicesKey from
+  // inInitializationCommMap_ and added it to devNCCLCommMap_
+  if (it != inInitializationCommMap_.end()) {
+    devNCCLCommMap_.emplace(devicesKey, std::move(it->second));
+    inInitializationCommMap_.erase(devicesKey);
+  }
+
+  it = devNCCLCommMap_.find(devicesKey);
+  TORCH_INTERNAL_ASSERT(
```

Comment:
This isn't the actual place where we would expect the `torch.distributed.DistBackendError` to be raised right, rather it would be `C10D_NCCL_CHECK_TIMEOUT_GROUPEND`, correct?

Happy to leave this assert here, just wondering if my intuition that there isn't a real way for it to fire is correct (if the function error'd out in the timeout check we wouldn't reach this assert).

Reply:
Yeah if this ASSERT fails it is a bug and hence I used `TORCH_INTERNAL_ASSERT`, which mentions that this is a bug that needs to be reported to PyTorch: https://github.com/pytorch/pytorch/blob/main/c10/util/Exception.h#L374